wire-library/wire-runtime-swift/src/main/swift/ProtoReader.swift

```diff
@@ -756,6 +797,38 @@ public final class ProtoReader {
         dictionary[unwrappedKey] = unwrappedValue
     }
 
+    private func handleUnknownEnumDictionaryEntry<T>(type: T.Type, startTag: UInt32, token: Int) throws  {
+        switch enumDecodingStrategy {
+        case .returnNil:
+            // We've found an unknown enum in a dictionary and need to write it to unknown fields.
+            // However, our reader has gone past the key and value, so we need to retrieve it.
+            // Example data:
+            // [0A] The tag that holds the map (1/Packed)
+            // [05] The length of the map (5)
+            // [0A] The tag of the key (1/Packed) <- We need the data starting here
+            // [01] The length of the key (1 - .HOME)
+            // [61] The value of the key: "a"
+            // [10] The tag of the value (2/Varint)
+            // [05] The value of the value (5 - unknown enum)
+            // [XX] <- Reader is now here
+            // We want to get the data for the last 5 of these bytes and write it to
+            // unknown fields in the parent tag, which will add back the first two bytes for us for us
+            let numberOfReadsPerKeyValuePair = 5
+
+            // NB: `buffer.getDataFromPastReads` does not move the pointer on the reader
+            // but creates a temporary pointer to get the historical data for us.
+            let keyValueData = try buffer.getDataFromPastReads(numPastReads: numberOfReadsPerKeyValuePair)
+
+            // It's important to end the message here, so that the unknown field is written at the parent
+            // message level (the first byte in the list above) and not our current nested level (the
+            // third byte in the list above).
+            _ = try endMessage(token: token)
+            try addUnknownField(tag: startTag, value: keyValueData)
+        case .throwError:
+            throw ProtoDecoder.Error.unknownEnumCase(type: T.self, fieldNumber: 1000)
+        }
+    }
+
```

Comment:
I'd need one of these for each type of key right?? The reason I didn't go down this route originally is because I didn't know the type for K and couldn't find a good way to be able to encode it without that. But yeah, if we want a specialized one for every type of [K: Int32] then that definitely removes the complexities I had to add to the ReadBuffer. 

 And thanks for the catch on the ordering. That makes sense since it's really just seeing a key/value pair as a message on its own, right? So the same rules apply to ordering not being guaranteed.

Reply:
Looking more into the code, I now see that for a given map [K:V] when V is RawRepresentable, we provide only one public method to decode into a dictionary, and K is a concrete type of String in that method. So it accepts a dict of [String: RawRepresentable].

Originally, I saw the neighboring public methods that accepted dicts such as [ProtoIntDecodable: ProtoDecodable] and [ProtoIntDecodable:  ProtoIntDecodable] and I think I assumed that we also had methods that accepted dicts like [ProtoIntDecodable, RawRepresentable]. But I see that's not the case now.

So does the protobuf spec not support a map of enums where the keys, for example, are Int32 and the values are an enums? Or does Wire just not support that case directly via its public methods? Or am I misunderstanding something? I'm pretty new to the internals of Protobuf here so that's very possibly the case :) 

For now going to move forward with the approach you suggested and support the [String: RawRepresentable] case directly.

Reply:
`[UInt32: Enum]` maps are supported...that seems to just be an oversight in Wire that we need to fix.

You're right that my example is a little too specific. You may end up needing a few different implementations of this shared `decode`, but hopefully some (if not all) of the ones where the value is an enum can be collapsed by using generics and potentially adding a new block parameter along the lines of `addUnknownPair: (UInt32, K, V) -> Void` (first param is the tag).

Hand waving a little bit, but if that doesn't work out let me know and we can discuss further.

Reply:
When you have a moment, PTAL at the latest commit. I think I was able to arrive at a solution that works well. I used your `addUnkownPair` closure idea to figure out the last mile -- so thanks for that idea! I backed out the ReadBuffer changes.

For now, I'm only calling the new generic `decode` method from the public [String: Enum] `decode`. But I'll be able to easily add a lightweight [ProtoIntDecodable: Enum] `decode` method (and any other missing ones) as well. Wasn't sure if you wanted me to add those to this PR or do it in a follow up. Thanks!